packages/subcommands/src/lib/utils/SubcommandsHandler.ts

```diff
@@ -0,0 +1,121 @@
+import type { Command } from '../structures/Command';
+import { SubcommandNotLoadedError } from '../errors/SubcommandNotLoadedError';
+import {
+	Args,
+	IPreconditionContainer,
+	PreCommandRunPayload,
+	SapphireClient,
+	Store,
+	PreconditionContainerArray,
+	PreconditionContainerSingle
+} from '@sapphire/framework';
+
+export type ResolveSubcommandPayload = Omit<PreCommandRunPayload, 'context'>;
+export interface ResolveSubcommandResult {
+	args: Args;
+	subcommand: Command | undefined;
+}
+
+export interface SubcommandsPluginOptions {
+	/**
+	 * Continue execution of parent command if command is disabled or error occurred
+	 * @since 1.0.0
+	 * @default false
+	 */
+	continueParent?: boolean;
+
+	/**
+	 * Run parent command when subcommand execution was ended
+	 * @since 1.0.0
+	 * @default false
+	 */
+	runParentAfterSubcommand?: boolean;
+
+	/**
+	 * Preconditions that will be overlap in parents by their subcommands
+	 * @since 1.0.0
+	 * @default []
+	 */
+	overlappedPreconditions?: string[];
+}
+
+export const enum SubcommandsEvents {
+	PreSubcommandRun = 'preSubcommandRun',
+	SubcommandDenied = 'subcommandDenied',
+	SubcommandAccepted = 'subcommandAccepted',
+	SubcommandRun = 'subcommandRun',
+	SubcommandSuccess = 'subcommandSuccess',
+	SubcommandError = 'subcommandError',
+	SubcommandFinish = 'subcommandFinish'
+}
+
+export class SubcommandsHandler {
+	// eslint-disable-next-line @typescript-eslint/naming-convention
+	public readonly subcommandsPluginOptions: Required<SubcommandsPluginOptions>;
+
+	public constructor(private readonly client: SapphireClient, options: SubcommandsPluginOptions) {
+		this.subcommandsPluginOptions = {
+			continueParent: options?.continueParent ?? false,
```

Comment:
Optional chaining is not necessary here since you're currently requiring `options` to be passed to the constructor. Either default `options` to `{}` and drop the optional chaining, or make it `options?: SubcommandPluginOptions`.